File: Sources/structlyC/05_binary_tree/tree_min_value.cpp
//
//  treeminvalue.cpp
//  MyCLI
//
//  Created by new on 3/25/26.
//

#include <algorithm>
#include <cmath>
#include <limits>

class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node(int initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

int treeMinValue(Node* root) {
    if (root == nullptr) {
        return std::numeric_limits<int>::infinity();
    }

    return std::min(std::min(root->val, treeMinValue(root->left)), treeMinValue(root->right));
}

File: Sources/structy/05_binary_tree/TreeMinValue.swift
//
//  TreeMinValue.swift
//  MyCLI
//
//  Created by new on 3/25/26.
//

func treeMinValue(_ root: TreeNode<Int>?) -> Int {
    let q = Queue<TreeNode<Int>>()
    if let root {
        q.enqueue(root)
    }
    var res = Int.max
    while !q.isEmpty, let node = q.dequeue() {
        res = min(res, node.val)
        if let left = node.left {
            q.enqueue(left)
        }
        if let right = node.right {
            q.enqueue(right)
        }
    }
    return res
}
